### PR TITLE
Typescript transition: ES6 import / export syntax 

### PR DESCRIPTION
**Pull request by cgewecke**
**NB: PR targets #144**
 
Work towards normalizing the way modules are loaded...

### CODE REVIEW

File: lib/service/fastethereumservice.ts [hunk 1]
```diff
@@ -25,7 +26,7 @@ export = module.exports = class FastEthereumService extends EthereumService {
    * @param {number}   [options.interval] sync retry interval
    * @param {Logger}   [options.logger] logger instance
    */
-  constructor (options: any) {
+  constructor (options?: any) {
```
Comment (holgerd77) [inline lib/service/fastethereumservice.ts:29]: Was this optionality of `options` necessary? I thought there is always passed an `options` dict around?

File: lib/service/service.ts [hunk 2]
```diff
@@ -12,15 +12,21 @@ const defaultOptions = {
...
  * @memberof module:service
  */
-export = module.exports = class Service extends EventEmitter {
+export class Service extends events.EventEmitter {
+  public logger: any
+  public opened: boolean
+  public running: boolean
+  public servers: any
+  public pool: any
+
```
Comment (holgerd77) [inline lib/service/service.ts:21]: Is this an intentional decision to make all these properties `public` on the various classes? Shouldn't we stay a bit more conservative here?

**Comment by holgerd77** [comment]
Phew, d4a0fd5597e2b7aebcb1b2749d32314ddd572fcc was a really tough one. I now transitioned the `Chain` class which lead to some deadlock problems, where I came to the conclusion that it is not worth to further dig into (after 2 hours of debugging *sigh* 😛 ) since I am pretty sure that this will be fixed in the upcoming blockchain release, see comment on tests.

`testdouble` problems remain. I also tried out `Sinon.js` and tried to do an alternative mock but this didn't went any better, had the impression that ES6 support in the library is rather worse than within `testdouble`. Ok, so we'll see. 

Didn't expect all this to be so hairy, grmblmbl. 🙃  Anyhow, think this will settle over time.

One note: remember to regularly test the CLI run as well (so `node dist/bin/cli.js`, after building), since otherwise the `bin/cli.ts` file is not tested and it is likely that this will need changes along transition work.

**Comment by holgerd77** [comment]
(won't further work on this today, so you can pick up again if you want)

File: lib/service/fastethereumservice.ts (hunk 1 above)
Comment (cgewecke) [reply lib/service/fastethereumservice.ts:29]: Did this in response to a compilation error...there was an empty constructor in the tests here...
https://github.com/ethereumjs/ethereumjs-client/blob/12ea4cdec363bc848f9fdf58c9094053b94f7a83/test/service/fastethereumservice.js#L26-L29

File: lib/service/service.ts (hunk 2 above)
Comment (cgewecke) [reply lib/service/service.ts:21]: Ah sorry!  Yes, agree that's a better default.

**Comment by cgewecke** [comment]
> Didn't expect all this to be so hairy, grmblmbl. 🙃 Anyhow, think this will settle over time.

Yes! It's really hairy...yesterday I saw the integration tests run a few tests and stop/exit without output for example. Commenting out code in unrelated test files allowed the run to progress further. Possibly `ts-node` compiling on the fly or "just in time" and crashing? lol/smh.

Will continue today 🙂

**Comment by cgewecke** [comment]
`net/peer` and `net/protocol` are converted. Commented out a few more tests, sadly. 

Have been leaving notes through the code in various places where `td` is broken or am not sure how to proceed. 

TS has a problem with `BoundProtocol`'s attempt to dynamically add properties to itself in `addMethods`
https://github.com/ethereumjs/ethereumjs-client/blob/515ce101f2ffbf44bfc579f095ac78624809fc6b/lib/net/protocol/boundprotocol.ts#L155-L159


This kind of redefinition happens in a few places in the tests as well. Have started casting those cases as below for now...  
```
(<unknown>Sender as any).someNewProperty
```

**Comment by holgerd77** [comment]
Ok, have worked on improving the Peer type integration, will stop for today. 😄

**Comment by holgerd77** [comment]
(trying to find a good balance here, but have the impression that some deepening in between makes sense to compensate a bit for the falling out tests)

**Comment by holgerd77** [comment]
Ok, will stop for today. Lots of changes, see commit messages for summary.

**Comment by cgewecke** [comment]
Summary:
+ updated @types/node to 14 (was getting a conflicting types error from other dependencies and this resolved it, idk)
+ swept the rest of the files replacing `require` with `import` (except `index.ts`)
+ coverage is working 
+ browser tests still failing :/ 

It looks like browser errors originate in `ethereumjs-block`. It's using `ethereumjs-util@v5.2.1` which uses `SafeBuffer` for `Buffer` like this:

```js
// ethereumjs-util/index.js
var Buffer = require('safe-buffer').Buffer;
```

Maybe there's some global namespace / buffer polyfill  conflict? 

A (slightly) cleaned-up stacktrace from the browser:
```
TypeError: Cannot read property 'from' of undefined
    at Object.global.wrappers./ethereumjs-client/node_modules/ethereumjs-util/dist/constants.js.buffer 
    at Object.global.wrappers./ethereumjs-client/node_modules/ethereumjs-util/dist/index.js../constants 
    at Object.global.wrappers./ethereumjs-client/node_modules/ethereumjs-block/index.js.ethereumjs-common 
    at Object.global.wrappers./ethereumjs-client/test/blockchain/chain.ts.ethereumjs-block 
    
 TypeError: Buffer is undefined
at ["/ethereumjs-client/node_modules/ethereumjs-util/dist/constants.js"]
require@http://localhost:9876/base/node_modules/karma-typescript/dist/client/commonjs.js
require/<@http://localhost:9876/base/node_modules/karma-typescript/dist/client/commonjs.js
global.wrappers["/ethereumjs-client/node_modules/ethereumjs-util/dist/index.js"]
require@http://localhost:9876/base/node_modules/karma-typescript/dist/client/commonjs.js
require/<@http://localhost:9876/base/node_modules/karma-typescript/dist/client/commonjs.js
global.wrappers["/ethereumjs-client/node_modules/ethereumjs-block/index.js"]
require@http://localhost:9876/base/node_modules/karma-typescript/dist/client/commonjs.js
require/<@http://localhost:9876/base/node_modules/karma-typescript/dist/client/commonjs.js
global.wrappers["/ethereumjs-client/test/blockchain/chain.ts"]
require@http://localhost:9876/base/node_modules/karma-typescript/dist/client/commonjs.js
```

Anyway...stopping here for today 🙂

**Comment by holgerd77** [comment]
@cgewecke Thanks, great 😄 , seems we are very much complete on the ES6 integration, so I would now merge this branch into the base TS transition branch.

**Pull request merged**